base/special/expm1.jl

```diff
@@ -0,0 +1,77 @@
+
+# min and max arguments by base and type
+MAX_EXP(::Type{Float64}) =  709.7827128933845   # log 2^1023*(2-2^-52)
+MIN_EXP(::Type{Float64}) = -37.42994775023705   # log 2^-54
+MAX_EXP(::Type{Float32}) =  88.72284f0          # log 2^127 *(2-2^-23)
+MIN_EXP(::Type{Float32}) = -17.32868f0          # log 2^-25
+
+# -ln(2) in upper and lower bits
+# Upper is truncated to only have 16 bits of significand since N has at most
+# ceil(log2(-MIN_EXP(n, Float32)*Ln2INV(Float32))) = 8 bits. (19 for Float64)
+# This ensures no rounding when multiplying Ln2U*N for FMAless hardware
+Ln2INV(::Type{Float32}) = 1.442695
+Ln2U(::Type{Float32}) = -0.69314575
+Ln2L(::Type{Float32}) = -1.4286068e-6
+Ln2INV(::Type{Float64}) = 1.4426950408889634
+Ln2U(::Type{Float64}) = -0.6931471805437468
+Ln2L(::Type{Float64}) = -1.619851018665656e-11
+
+@inline function exthorner(x, p::Tuple)
+    # polynomial evaluation using compensated summation.
+    # much more accurate, especially when lo can be combined with other rounding errors
+    hi, lo = p[end], zero(x)
+    for i in length(p)-1:-1:1
+        pi = p[i]
+        prod = hi*x
+        err = fma(hi, x, -prod)
+        hi = pi+prod
+        lo = fma(lo, x, prod - (hi - pi) + err)
+    end
+    return hi, lo
+end
+
+@inline function expm1_kernel(x::Float64)
+    hi_order = evalpoly(x, (0.1666666666666668, 0.041666666666666706, 0.0083333333333196,
+                            0.0013888888888864692, 0.00019841269890049138, 2.4801587363805883e-5,
+                            2.7557240916652906e-6, 2.755724050237237e-7, 2.511003824738655e-8,
+                            2.092503179646227e-9))
+    return exthorner(x, (1.0, 0.5, hi_order))
+end
+
+
+@doc """
+    expm1(x)
+
+Accurately compute ``e^x-1``.
+"""
+expm1(x::Real) = expm1(float(x))
+@inline function expm1(x::T) where T<:Float64
+    N_float = round(x*Ln2INV(T))
+    N = unsafe_trunc(uinttype(T), N_float)
+    r = muladd(N_float, Ln2U(T), x)
+    r = muladd(N_float, Ln2L(T), r)
+    hi, lo = expm1_kernel(r)
+    small_part = r*hi
+    small_round = fma(r, lo, fma(r, hi, -small_part))
+    twopk = reinterpret(T, (N+uinttype(T)(exponent_bias(T))) << uinttype(T)(significand_bits(T)))
+    
+    if !(abs(x)<=MIN_EXP(T))
+        isnan(x) && return x
+        x > MAX_EXP(T) && return T(Inf)
+        x < MIN_EXP(T) && return T(-1)
+        if N == exponent_max(T)
+            return muladd(small_part, T(2), T(2)) * T(2)^(exponent_max(T)-1)
+        end
+    end
+    return fma(twopk, small_round, fma(twopk, small_part, twopk-T(1)))
+end
+@inline function myexpm1(x::T) where T<:Float32
```

Comment:
Is this a test function or intended to be in here?

Reply:
It's intended at least for now. It looks like a very good methods that won't be easy to make generic, so my current plan for this is to write a fully starts Float64 version.

Reply:
@oscardssmith, I don't understand your second sentence…

Reply:
That's probably cause it made no sense.